curl_cffi/requests/session.py

```diff
@@ -335,21 +335,27 @@ def _set_curl_options(
         # proxies
         if self.proxies:
             proxies = {**self.proxies, **(proxies or {})}
+
         if proxies:
-            if url.startswith("http://"):
-                if proxies["http"] is not None:
-                    c.setopt(CurlOpt.PROXY, proxies["http"])
-            elif url.startswith("https://"):
-                if proxies["https"] is not None:
-                    if proxies["https"].startswith("https://"):
-                        raise RequestsError(
-                            "You are using http proxy WRONG, the prefix should be 'http://' not 'https://',"
-                            "see: https://github.com/yifeikong/curl_cffi/issues/6"
-                        )
-                    c.setopt(CurlOpt.PROXY, proxies["https"])
-                    # for http proxy, need to tell curl to enable tunneling
-                    if not proxies["https"].startswith("socks"):
-                        c.setopt(CurlOpt.HTTPPROXYTUNNEL, 1)
+            parts = urlparse(url)
+            proxy = proxies.get(parts.scheme, proxies.get("all"))
+            if parts.hostname:
+                proxy = proxies.get(
+                    f"{parts.scheme}://{parts.hostname}",
+                    proxies.get(f"all://{parts.hostname}"),
+                ) or proxy
+
+            if proxy is not None:
+                if parts.scheme == "https" and proxy.startswith("https://"):
+                    raise RequestsError(
+                        "You are using http proxy WRONG, the prefix should be 'http://' not 'https://',"
+                        "see: https://github.com/yifeikong/curl_cffi/issues/6"
+                    )
```

Comment:
In fact, curl supports HTTPS proxies [since v7.52](https://daniel.haxx.se/blog/2016/11/26/https-proxy-with-curl/).

Could it possible to support it in `curl-cffi`?
Also, recently `httpcore` (core of `httpx`): https://github.com/encode/httpcore/pull/745

Reply:
Yes, I'm aware of that. However, in real worlds, no one uses TLS-in-TLS proxies, they just mistakenly think that http proxy for https urls should start with `https://`.

Reply:
> they just mistakenly think that http proxy for https urls should start with `https://`.

Then we could show Error only when we are sure it used mistakenly.

Instead of raise error before making requests:
```py
(CurlError("Failed to perform, ErrCode: 35, Reason: 'error:100000f7:SSL routines:OPENSSL_internal:WRONG_VERSION_NUMBER'"),)
```

We could wrap CurlError if proxy scheme is HTTPS, then show additional warning:
```py
     warning.warn(
        "You are using httpS proxy, we noticed proxy-server doesn't support HTTPS,"
        " try replace proxy protocol 'https://' with 'http://' and check if it's helping."
     )
```

Reply:
NordVPN proxies use https, I use them sometimes.